core/domain/stats_domain.py

```diff
@@ -16,213 +16,250 @@
+
+        if self.rule_spec_str is not None and not isinstance(
+                self.rule_spec_str, basestring):
+            raise utils.ValidationError(
+                'Expected rule_spec_str to be either None or a string, '
+                'received %s' % str(self.rule_spec_str))
+
+        if self.answer_str is not None and not isinstance(
+                self.answer_str, basestring):
+            raise utils.ValidationError(
+                'Expected answer_str to be either None or a string, received '
+                '%s' % str(self.answer_str))
+
+        if not isinstance(self.session_id, basestring):
+            raise utils.ValidationError(
+                'Expected session_id to be a string, received %s' %
+                str(self.session_id))
+
+        if not isinstance(self.time_spent_in_sec, float):
             raise utils.ValidationError(
-                'Expected answers_list to be a list, received %s' %
-                self.answers_list)
+                'Expected time_spent_in_sec to be a float, received %s' %
+                str(self.time_spent_in_sec))
 
-        # Note: There is no need to validate content of answers_list here
-        # because each answer is validated before it is appended to
-        # answers_list (which is faster than validating whole answers_list
-        # each time a new answer is recorded).
+        if not isinstance(self.params, dict):
+            raise utils.ValidationError(
+                'Expected params to be a dict, received %s' % str(self.params))
+
+        if not isinstance(self.answer_group_index, int):
+            raise utils.ValidationError(
+                'Expected answer_group_index to be an integer, received %s' %
+                str(self.answer_group_index))
+
+        if not isinstance(self.rule_spec_index, int):
+            raise utils.ValidationError(
+                'Expected rule_spec_index to be an integer, received %s' %
+                str(self.rule_spec_index))
+
+        if not isinstance(self.classification_categorization, basestring):
+            raise utils.ValidationError(
+                'Expected classification_categorization to be a string, '
+                'received %s' % str(self.classification_categorization))
+
+        if self.answer_group_index < 0:
+            raise utils.ValidationError(
+                'Expected answer_group_index to be non-negative, received %d' %
+                self.answer_group_index)
+
+        if self.rule_spec_index < 0:
+            raise utils.ValidationError(
+                'Expected rule_spec_index to be non-negative, received %d' %
+                self.rule_spec_index)
+
+        if self.time_spent_in_sec < 0.:
+            raise utils.ValidationError(
+                'Expected time_spent_in_sec to be non-negative, received %f' %
+                self.time_spent_in_sec)
+
+        if self.normalized_answer is None and self.interaction_id != 'Continue':
```

Comment:
(Note that interaction.is_linear also exists, if you want to use that. May be more trouble than it's worth, though -- maybe just make a list and have a backend test to ensure that that list always reflects reality.)